### PR TITLE
Adding num function/grad/prox/ls eval in solvers

### PR DESCRIPTION
**Pull request by zaccharieramzi**
For now HZ-LS does not have the right elements in the state so it will fail in all solvers

Added num function/grad/ls eval in the following solvers:
- (L)-BFGS-(B): added num fun/grad/ls
- Anderson: added num fun
- Armijo SGD: for now armijo LS does not return the correct elements so difficult to do, esp. because the loops do not return the number of iterations done
- Bisection: added num fun
- BlockCD: added num fun/prox/grad
- Broyden: added num fun/ls
- FixedPointIteration: added num fun
- Gauss-Newton: not doable, because at some point linear solve does not return the number of iterations done. This comes down to `_cg_solve` returning only the final result.
- (Proximal/Projected)GradientDescent: the fista LS does not return the number of iterations done (same problem as Armijo)
- IterativeRefinement: added num matvec/matvec bar/solve
- LevenbergMarquardt: not sufficient info returned by the solver (+ extremely complex code)
- MirrorDescent: added num fun/grad/proj
- NonLinearCG: added num fun/grad/ls
- OptaxWrapper: I don't think we have this info in every optax state, which in any case can be looked up with the current API
- PolyakSGD: added num fun/grad
- ScipyWrapper(s): added num fun/jac/hess in scipy minimize, added num fun (and num iter) in scipy root

### CODE REVIEW

**Comment by zaccharieramzi** [comment]
@mblondel This is ready for review. 
A couple of comments:
- GaussNewton suffers from an inefficiency in function calls -> will tackle in a subsequent PR
- HZ-LS does not have num fun/grad evals
- Backtracking LS for now does not report num grad eval correctly (it is underestimated), but hopefully, the inefficiency will be corrected and this will be consistent
- Some adhoc line searches (armijo, fista) do not report the num of evals because they use `_while_loop` which does not report the number of iterations -> maybe they should be framed with the new paradigm of @vroulet @fabianp

**Comment by zaccharieramzi** [comment]
Actually, I think that adding a unit test at least for num fun eval would be welcome, writing one

**Comment by zaccharieramzi** [comment]
@mblondel this is re-ready for review: the unit test is written and passes, HZ LS now has the num function eval

File: jaxopt/_src/armijo_sgd.py [hunk 1]
```diff
@@ -127,6 +127,9 @@ class ArmijoState(NamedTuple):
   aux: Optional[Any]
   stepsize: float
   velocity: Optional[Any]
+  num_fun_eval: jnp.array = jnp.array(0)
+  num_grad_eval: jnp.array = jnp.array(0)
+  num_line_search_iter: jnp.array = jnp.array(0)
```
Comment (mblondel) [inline jaxopt/_src/armijo_sgd.py:132]:
```suggestion
  num_linesearch_iter: jnp.array = jnp.array(0)
```
for consistency

File: jaxopt/_src/armijo_sgd.py [hunk 2]
```diff
@@ -281,12 +287,16 @@ def update(self, params, state, *args, **kwargs) -> base.OptStep:
...
+    new_num_fun_eval = state.num_fun_eval + 1
+    new_num_grad_eval = state.num_grad_eval + 1
 
     goldstein = self.reset_option == 'goldstein'
     stepsize, next_params, f_next = self._armijo_line_search(
       goldstein, self.maxls, params, f_cur, stepsize, grad, self._coef,
       self.decrease_factor, self.increase_factor, self.max_stepsize, args,
       kwargs)
+    # FIXME: update new_num_fun_eval, new_num_grad_eval, new_num_line_search_iter
+    # with outputs from armijo line search
```
Comment (mblondel) [inline jaxopt/_src/armijo_sgd.py:299]: Does that mean that the current returned numbers are incorrect?

File: jaxopt/_src/armijo_sgd.py (hunk 1 above)
Comment (mblondel) [reply jaxopt/_src/armijo_sgd.py:132]: and same everywhere

File: tests/common_test.py [hunk 3]
```diff
@@ -96,6 +102,60 @@ def fixed_point_fun(params, *args, **kwargs):
...
+    common_kwargs = dict(jit=False, maxiter=30)
+    fpi = jaxopt.FixedPointIteration(fixed_point_fun=fixed_point_fun, **common_kwargs)
+
+    # Gradient solvers and fixed point
+    for solver in (jaxopt.PolyakSGD(fun=fun, **common_kwargs),
+                   fpi,
+                   jaxopt.Broyden(fun=root_fun, **common_kwargs),
+                   jaxopt.AndersonAcceleration(fixed_point_fun=fixed_point_fun, **common_kwargs),
+                   jaxopt.BFGS(fun, **common_kwargs),
+                   jaxopt.LBFGS(fun=fun, **common_kwargs)):
```
Comment (mblondel) [inline tests/common_test.py:134]: Can you add more of them here?

File: jaxopt/_src/armijo_sgd.py (hunk 2 above)
Comment (zaccharieramzi) [reply jaxopt/_src/armijo_sgd.py:299]: Yes, this is a problem.

I will remove these values for the returned state, and add an issue specifying which solvers need to be updated to feature these values.

File: tests/common_test.py (hunk 3 above)
Comment (zaccharieramzi) [reply tests/common_test.py:152]: which ones were you thinking about? For the (prox/proj) gradient descent ones because of the current fista line search it is not possible to have the right results.

I added nonlinear CG

Comment (zaccharieramzi) [reply tests/common_test.py:152]: (which allowed to correct a zoom LS related bug)

File: jaxopt/_src/armijo_sgd.py (hunk 1 above)
Comment (zaccharieramzi) [reply jaxopt/_src/armijo_sgd.py]: done

File: tests/common_test.py (hunk 3 above)
Comment (mblondel) [reply tests/common_test.py:152]: All the ones for which you added support for `num_fun_evals`.

Comment (zaccharieramzi) [reply tests/common_test.py:176]: Sure : added MirrorDescent, but it's quite a big one

For BlockCD I can't add it because BlockCD does not support non-jitted functions yet (so it's not possible to build the unit test with the global variable like I did). I added a fixme, and will report this in an issue (should be a similar resolution as what @vroulet did with the cond in the LS).

**Comment by mblondel** [comment]
LGTM, apart from squashing.

**Comment by zaccharieramzi** [comment]
I actually need to correct the tests that are now red after the weak type consistency checks. I am planning to force the dtype to be jnp.int32 in all num evals, wdyt?

**Comment by mblondel** [comment]
Yes I think this should be enough
```python
>>> jnp.iinfo(jnp.int32).max
2147483647
```

**Comment by zaccharieramzi** [comment]
@mblondel corrected the dtypes, can you make sure it's ok before I squash the commits?

**Comment by mblondel** [comment]
Is there a rational for `iter_num` being `int64` and `num_grad_eval` int32`?

File: jaxopt/_src/anderson.py [hunk 4]
```diff
@@ -108,6 +108,8 @@ class AndersonState(NamedTuple):
   residual_gram: jnp.array
   aux: Any
 
+  num_fun_eval: jnp.array = jnp.array(0, base.N_EVAL_DTYPE)
```
Comment (mblondel) [inline jaxopt/_src/anderson.py:111]: Sorry, slight preference for `NUM_EVAL_DTYPE` for consistency

Comment (zaccharieramzi) [reply jaxopt/_src/anderson.py:111]: sure, fixing rn

File: jaxopt/_src/base.py [hunk 5]
```diff
@@ -56,7 +56,7 @@ class KKTSolution(NamedTuple):
   dual_ineq: Optional[Any] = None
 
 
-N_EVAL_DTYPE = jnp.int64
+NUM_EVAL_DTYPE = jnp.int64
```
Comment (mblondel) [inline jaxopt/_src/base.py:59]: Sorry I would prefer if we use int32. Should hopefully be my last comment :)

Comment (mblondel) [reply jaxopt/_src/base.py:59]: The reason is because I'm worried int64 could be problematic on TPU. Plus, int32 should be enough in most use cases.

**Comment by mblondel** [comment]
Ready for squashing, thanks!

File: jaxopt/_src/base.py (hunk 5 above)
Comment (zaccharieramzi) [reply jaxopt/_src/base.py]: makes sense, reverted to int64
